Building CUDA code in mickledore SDK fails due to "-mbranch-protection=standard" flag

I encountered an issue trying to build CUDA code in the SDK of our yocto image based on mickledore. The issue seems to be related to #1302.

When I try to configure a cmake project with includes CUDA code, I get the following error message:

> ...
> CMake Error at /opt/sdk/sysroots/x86_64-pyritesdk-linux/usr/share/cmake-3.25/Modules/CMakeDetermineCompilerId.cmake:739 (message):
>   Compiling the CUDA compiler identification source file
>   "CMakeCUDACompilerId.cu" failed.
> 
>   Compiler:
>   /opt/sdk/sysroots/x86_64-pyritesdk-linux/usr/local/cuda-11.4/bin/nvcc
> 
>   Build flags:
>   --gpu-architecture=compute_87;--gpu-code=sm_87;--include-path;/opt/sdk/sysroots/armv8a-leica-linux/usr/local/cuda-11.4/include;--library-path;/opt/sdk/sysroots/armv8a-leica-linux/usr/local/cuda-11.4/lib;-Xcompiler;--sysroot=/opt/sdk/sysroots/armv8a-leica-linux;-Xcompiler;-march=armv8-a+crc,-mbranch-protection=standard,-fstack-protector-strong,-O2,-D_FORTIFY_SOURCE=2,-Wformat,-Wformat-security,-Werror=format-security
> 
> ...
> 
> #$
>   "/opt/sdk/sysroots/x86_64-pyritesdk-linux/usr/bin/aarch64-leica-linux"/aarch64-leica-linux-g++-8.5.0
>   -D__CUDA_ARCH__=870 -E -x c++ -DCUDA_DOUBLE_MATH_FUNCTIONS -D__CUDACC__
>   -D__NVCC__ --sysroot=/opt/sdk/sysroots/armv8a-leica-linux
>   -march=armv8-a+crc -mbranch-protection=standard -fstack-protector-strong
>   -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security
>   -I"/opt/sdk/sysroots/armv8a-leica-linux/usr/local/cuda-11.4/include"
>   "-I/opt/sdk/sysroots/x86_64-pyritesdk-linux/usr/local/cuda-11.4/bin/..//include"
>   -D__CUDACC_VER_MAJOR__=11 -D__CUDACC_VER_MINOR__=4
>   -D__CUDACC_VER_BUILD__=315 -D__CUDA_API_VER_MAJOR__=11
>   -D__CUDA_API_VER_MINOR__=4 -include "cuda_runtime.h"
>   "CMakeCUDACompilerId.cu" -o "tmp/CMakeCUDACompilerId.cpp1.ii"
> 
>   aarch64-leica-linux-g++-8.5.0: error: unrecognized command line option
>   ‘-mbranch-protection=standard’; did you mean
>   ‘-fexcess-precision=standard’?
> 
>   # --error 0x1 --
> 
> ...

The build flag "-mbranch-protection=standard" seems to cause this problem. This flag has recently been removed from _gcc-configure-common.inc_ (see #1302).

In the SDK, the build flag is defined in _$SDKTARGETSYSROOT/environment-setup.d/cuda_target.sh_:

> export CUDA_NVCC_ARCH_FLAGS="--gpu-architecture=compute_87 --gpu-code=sm_87"
> export CUDA_ARCHITECTURES="87"
> export CUDACXXARCHFLAGS="-Xcompiler -march=armv8-a+crc,-mbranch-protection=standard,-fstack-protector-strong,-O2,-D_FORTIFY_SOURCE=2,-Wformat,-Wformat-security,-Werror=format-security"
> export CUDAHOSTCXX="aarch64-leica-linux-g++-8.5.0"

This file is generated by the _cuda-target-environment_1.0.bb_ recipe:

> 
> ...
> 
> inherit cuda-gcc
> 
> ...
> 
> def arch_flags(d):
>     archflags = d.getVar('TARGET_CC_ARCH')
>     if archflags:
>         return "-Xcompiler " + ','.join(archflags.split())
>     return ""
> 
> do_compile() {
>     sed -e"s!@CUDA_NVCC_ARCH_FLAGS@!${CUDA_NVCC_ARCH_FLAGS}!" \
> 	-e"s!@ARCHFLAGS@!${@arch_flags(d)}!" \
> 	-e"s!@CUDA_ARCHITECTURES@!${CMAKE_CUDA_ARCHITECTURES}!" \
> 	-e"s!@COMPILER_CMD@!${COMPILER_CMD}!" ${S}/cuda_target.sh.in > ${B}/cuda_target.sh
> }
> 
> ...
> 

The TARGET_CC_ARCH variable is set to TUNE_CCARGS in _poky/meta/conf/bitbake.conf_.

The fix implemented in #1306 removes the -mbranch-protection=standard flag from TUNE_CCARGS in the _cuda.bbclass_, but not in the _cuda-gcc.bbclass_ which is inherited in _cuda-target-environment_1.0.bb_.
Therefore, the cuda_target.sh file is still generted with the -mbranch-protection=standard flag.

I would propose moving the fix implemented in #1306
`TUNE_CCARGS:remove:cuda = "-mbranch-protection=standard"`
from the _cuda.bbclass_ class to the _cuda-gcc.bbclass_ class, since the cuda class also inherits cuda-gcc.
Alternatively, the flag could also be removed in the _cuda-target-environment_1.0.bb_ recipe directly.

Thanks for your help!
Cheers,
Patrick

--- a/recipes-devtools/cuda/cuda-target-environment_1.0.bb
+++ b/recipes-devtools/cuda/cuda-target-environment_1.0.bb
@@ -14,9 +14,9 @@ COMPILER_CMD  = "${@d.getVar('CXX_FOR_CUDA').split()[0]}"
 CMAKE_CUDA_ARCHITECTURES = "${@d.getVar('CUDA_ARCHITECTURES') if d.getVar('CUDA_ARCHITECTURES') else 'OFF'}"
 
 def arch_flags(d):
-    archflags = d.getVar('TARGET_CC_ARCH')
+    archflags = [flag for flag in (d.getVar('TARGET_CC_ARCH') or '').split() if not flag.startswith('-mbranch-protection')]
     if archflags:
-        return "-Xcompiler " + ','.join(archflags.split())
+        return "-Xcompiler " + ','.join(archflags)
     return ""
 
 do_compile() {

